bee-tangle/src/traversal.rs

```diff
@@ -0,0 +1,170 @@
+// /// associated data and metadata.
+// pub async fn visit_children_follow_first_parent<Metadata, Match, Apply, H: Hooks<Metadata>>(
+//     tangle: &Tangle<Metadata, H>,
+//     root: MessageId,
+//     mut matches: Match,
+//     mut apply: Apply,
+// ) where
+//     Metadata: Clone + Copy,
+//     Match: FnMut(&MessageRef, &Metadata) -> bool,
+//     Apply: FnMut(&MessageId, &MessageRef, &Metadata),
+// {
+//     // TODO could be simplified like visit_parents_follow_parent1 ? Meaning no vector ?
+//     let mut children = vec![root];
+//
+//     while let Some(ref parent_id) = children.pop() {
+//         if let Some(parent) = tangle.get_vertex(parent_id).await {
+//             if matches(parent.message(), parent.metadata()) {
+//                 apply(parent_id, parent.message(), parent.metadata());
+//
+//                 if let Some(parent_children) = tangle.get_children(parent_id).await {
+//                     for child_id in parent_children {
+//                         if let Some(child) = tangle.get_vertex(&child_id).await {
+//                             if &child.parents()[0] == parent_id {
+//                                 children.push(child_id);
+//                             }
+//                         }
+//                     }
+//                 }
+//             }
+//         }
+//     }
+// }
+
+/// A Tangle walker that - given a starting vertex - visits all of its ancestors that are connected through
+/// either the *parent1* or the *parent2* edge. The walk continues as long as the visited vertices match a certain
+/// condition. For each visited vertex customized logic can be applied depending on the availability of the
+/// vertex. Each traversed vertex provides read access to its associated data and metadata.
+pub async fn visit_parents_depth_first<Fut, Metadata, Match, Apply, ElseApply, MissingApply, H: Hooks<Metadata>>(
+    tangle: &Tangle<Metadata, H>,
+    root: MessageId,
+    matches: Match,
+    mut apply: Apply,
+    mut else_apply: ElseApply,
+    mut missing_apply: MissingApply,
+) where
+    Fut: Future<Output = bool>,
+    Metadata: Clone + Copy,
+    Match: Fn(MessageId, MessageRef, Metadata) -> Fut,
+    Apply: FnMut(&MessageId, &MessageRef, &Metadata),
+    ElseApply: FnMut(&MessageId, &MessageRef, &Metadata),
+    MissingApply: FnMut(&MessageId),
+{
+    let mut parents = Vec::new();
+    let mut visited = HashSet::new();
+
+    parents.push(root);
+
+    while let Some(message_id) = parents.pop() {
+        if !visited.contains(&message_id) {
+            let msg_meta = tangle
+                .get_vertex(&message_id)
+                .await
+                .as_ref()
+                .and_then(|v| v.message_and_metadata().cloned());
+            match msg_meta {
+                Some((msg, meta)) => {
+                    if matches(message_id, msg.clone(), meta).await {
+                        apply(&message_id, &msg, &meta);
+
+                        for parent in msg.parents().iter() {
```

Comment:
could we use `extend_from_slice` or preallocate space in `parents` to avoid multiple allocation calls?

Reply:
Now that `Parents` has been changed, yes